nativescript-angular/router/ns-location-strategy.ts

```diff
@@ -432,13 +393,246 @@ export class NSLocationStrategy extends LocationStrategy {
             animated: isPresent(options.animated) ? options.animated : true,
             transition: options.transition
         };
+
         if (isLogEnabled()) {
             routerLog("NSLocationStrategy._setNavigationOptions(" +
-            `${JSON.stringify(this._currentNavigationOptions)})`);
+                `${JSON.stringify(this._currentNavigationOptions)})`);
         }
     }
 
-    public _getStates(): { [key: string]: Array<LocationState> } {
-        return this.statesByOutlet;
+    public _getOutlets(): Array<Outlet> {
+        return this.outlets;
+    }
+
+    updateOutletFrame(outlet: Outlet, frame: Frame) {
+        outlet.frame = frame;
+        this.currentOutlet = outlet;
+    }
+
+    clearOutlet(frame: Frame) {
+        this.outlets = this.outlets.filter(currentOutlet => {
+            // Remove current outlet, if it's named p-r-o, from the url tree.
+            const isEqualToCurrent = currentOutlet.pathByOutlets === this.currentOutlet.pathByOutlets;
+            const isModalOutlet = currentOutlet.modalNavigationDepth > 0;
+
+            if (currentOutlet.frame === frame && isModalOutlet && !isEqualToCurrent) {
+                this.callPopState(null, true, currentOutlet);
+            }
+            return currentOutlet.frame !== frame;
+        });
+    }
+
+    getSegmentGroupFullPath(segmentGroup: UrlSegmentGroup): string {
+        let fullPath = "";
+
+        while (segmentGroup) {
+            const url = segmentGroup.toString();
+            fullPath = fullPath ? (url ? url + "/" : url) + fullPath : url;
```

Comment:
This one-liner is concise but that also makes it unnecessarily complex to understand.